base/compiler/abstractinterpretation.jl

```diff
@@ -1811,7 +1811,12 @@ function abstract_eval_special_value(interp::AbstractInterpreter, @nospecialize(
     elseif isa(e, SSAValue)
         return abstract_eval_ssavalue(e, sv)
     elseif isa(e, SlotNumber) || isa(e, Argument)
-        return vtypes[slot_id(e)].typ
+        sn = slot_id(e)
+        s = vtypes[sn]
+        if s.undef
+            sv.src.slotflags[sn] |= SLOT_USEDUNDEF | SLOT_STATICUNDEF
+        end
```

Comment:
After digging into the regressions in the optimizer, I found this change could be problematic, and especially causes some regressions in the SROA pass.

I think these changes are orthogonal to the main purpose of this PR, so I am trying to fix regressions by separating related changes into another branch.

Still there seems to be another cause of the regressions though.